rest/diagnostic_doc_api_test.go

```diff
@@ -57,3 +61,150 @@ func TestGetAlldocChannels(t *testing.T) {
+		bucket, base.TestUseXattrs(), SyncFn, ImportFilter))
+	RequireStatus(t, resp, http.StatusCreated)
+	response := rt.SendDiagnosticRequest("GET", "/{{.keyspace}}/_sync", `{"accessChannel": ["dynamicChan5412"],"accessUser": "user","channel": ["dynamicChan222"],"expiry":10}`)
+	RequireStatus(t, response, http.StatusOK)
+	var respMap SyncFnDryRun
+	var respMapinit SyncFnDryRun
+	err := json.Unmarshal(response.BodyBytes(), &respMapinit)
+	assert.NoError(t, err)
+	assert.ElementsMatch(t, respMapinit.Exception, nil)
+	assert.Equal(t, respMapinit.Roles, channels.AccessMap{})
+	assert.Equal(t, respMapinit.Access, channels.AccessMap{"user": channels.BaseSetOf(t, "dynamicChan5412")})
+	assert.EqualValues(t, *respMapinit.Expiry, 10)
+	response = rt.SendDiagnosticRequest("GET", "/{{.keyspace}}/_sync", `{"role": ["role:role1"], "accessUser": "user"}`)
+	RequireStatus(t, response, http.StatusOK)
+
+	err = json.Unmarshal(response.BodyBytes(), &respMap)
+	assert.NoError(t, err)
+	assert.Equal(t, respMap.Roles, channels.AccessMap{"user": channels.BaseSetOf(t, "role1")})
+	newSyncFn := `"function(doc,oldDoc){if (doc.user.num >= 100) {channel(doc.channel);} else {throw({forbidden: 'user num too low'});}if (oldDoc){ console.log(oldDoc); if (oldDoc.user.num > doc.user.num) { access(oldDoc.user.name, doc.channel);} else {access(doc.user.name[0], doc.channel);}}}"`
+	resp = rt.SendAdminRequest("POST", "/db/_config", fmt.Sprintf(
+		`{"bucket":"%s", "num_index_replicas": 0, "enable_shared_bucket_access": %t, "sync":%s, "import_filter":%s}`,
+		bucket, base.TestUseXattrs(), newSyncFn, ImportFilter))
+	RequireStatus(t, resp, http.StatusCreated)
+
+	_ = rt.PutDoc("doc", `{"user":{"num":123, "name":["user1"]}, "channel":"channel1"}`)
+
+	response = rt.SendDiagnosticRequest("GET", "/{{.keyspace}}/_sync", `{"user":{"num":23}}`)
+	RequireStatus(t, response, http.StatusOK)
+
+	err = json.Unmarshal(response.BodyBytes(), &respMap)
+	assert.NoError(t, err)
+	assert.Equal(t, respMap.Exception, "403 user num too low")
+
+	response = rt.SendDiagnosticRequest("GET", fmt.Sprintf("/{{.keyspace}}/_sync?doc_id=doc"), `{"user":{"num":150}, "channel":"abc"}`)
+	RequireStatus(t, response, http.StatusOK)
+	var newrespMap SyncFnDryRun
+	err = json.Unmarshal(response.BodyBytes(), &newrespMap)
+	assert.NoError(t, err)
+	assert.Equal(t, newrespMap.Exception, "TypeError: Cannot access member '0' of undefined")
+
+	response = rt.SendDiagnosticRequest("GET", fmt.Sprintf("/{{.keyspace}}/_sync?doc_id=doc"), `{"user":{"num":120, "name":["user2"]}, "channel":"channel2"}`)
+	RequireStatus(t, response, http.StatusOK)
+
+	err = json.Unmarshal(response.BodyBytes(), &respMap)
+	assert.NoError(t, err)
+	assert.Equal(t, respMap.Access, channels.AccessMap{"user1": channels.BaseSetOf(t, "channel2")})
+
+	// get doc from bucket with no body provided
+	response = rt.SendDiagnosticRequest("GET", fmt.Sprintf("/{{.keyspace}}/_sync?doc_id=doc"), ``)
```

Comment:
This case is using the document from the bucket as *both* doc and oldDoc in the sync function.  Is this intentional?    Using it as doc only seems like a more natural use case to me, but I'm open to other points of view.

Reply:
I remember we discussed allowing this in sync up, either we use it as doc only or we attempt to get the previous rev as oldDoc?

Reply:
Ive made the change to use it as doc only, let me know if we need to pull the old rev as old doc